sshd-common/src/main/java/org/apache/sshd/common/cipher/BaseGCMCipher.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.sshd.common.cipher;
+
+import java.security.spec.AlgorithmParameterSpec;
+
+import javax.crypto.Cipher;
+import javax.crypto.spec.GCMParameterSpec;
+
+import org.apache.sshd.common.util.buffer.Buffer;
+import org.apache.sshd.common.util.buffer.ByteArrayBuffer;
+
+public class BaseGCMCipher extends BaseAEADCipher {
+
+    public BaseGCMCipher(
+                         int ivsize, int authSize, int kdfSize, String algorithm, int keySize, String transformation,
+                         int blkSize) {
+        super(ivsize, authSize, kdfSize, algorithm, keySize, transformation, blkSize);
+    }
+
+    @Override
+    protected AlgorithmParameterSpec initializeAlgorithmParameters(byte[] iv) {
+        Buffer buffer = new ByteArrayBuffer(iv);
+        buffer.rpos(Integer.BYTES);
+        long ic = buffer.getLong();
+        // decrement IV as it will be incremented back to initial value on first call to updateWithAAD
+        ic = (ic - 1) & 0x0ffffffffL;
+        buffer.wpos(Integer.BYTES);
+        buffer.putLong(ic);
+        return new GCMParameterSpec(getAuthenticationTagSize() * Byte.SIZE, buffer.array());
+    }
+
+    @Override
+    protected AlgorithmParameterSpec getNextAlgorithmParameters() {
+        Cipher cipher = getCipherInstance();
+        Buffer iv = new ByteArrayBuffer(cipher.getIV());
+        iv.rpos(Integer.BYTES);
+        long ic = iv.getLong();
+        ic = (ic + 1) & 0x0ffffffffL;
+        iv.wpos(Integer.BYTES);
+        iv.putLong(ic);
+        return new GCMParameterSpec(getAuthenticationTagSize() * Byte.SIZE, iv.array());
+    }
```

Comment:
Pretty expensive in memory - every time it is called (which is for every packet) a new `byte[]` (`cipher.getIv()` returns a **clone**), new `Buffer` and new `GCMParameterSpec` instance is created. Can we do something about it ?

Moreover, this code is exactly the same as `initializeAlgorithmParameters` except it does `+1` instead of `-1` - perhaps we could use an internal (protected) `updateAlgorithmParameters(byte[] iv, long delta)` and call it once with `+1` and once with `-1`

Reply:
I believe we can avoid creating the `Buffer` since all the code does is replace an 8-byte section of the IV at offset 4:

```java
byte[] iv = cipher.getIV();
// Need to write decodeLong  but it should be similar to decodeInt in the same class
// can use the code of Buffer#getLong as base
long ic = KeyEntryResolver.decodeLong(iv, Integer.BYTES, iv.length - Integer.BYTES);
ic = (ic + delta) & 0x0ffffffffL;
// Need to write it - can use the code of Buffer#putLong as base
KeyEntryResolver.encodeLong(iv, ic, Integer.BYTES, iv.length - Integer.BYTES);
```

Reply:
Ooh, nice find. Yes, we just need to update the 8-byte section of the buffer. The update algorithm parameters API idea sounds good to me; it's more in line with how OpenSSH is structured, too.

Reply:
Regarding the `GCMParameterSpec` based on [this link](https://alvinalexander.com/java/jwarehouse/openjdk-8/jdk/src/share/classes/javax/crypto/spec/GCMParameterSpec.java.shtml) we might be able to execute an ugly hack via reflection API and manipulate the internal IV data directly instead of creating a new instance every time... will need to think about it some more - but only **after** this code is working.

Reply:
I came up with a better idea: extend the class and don't clone the IV. Less safe, but doesn't continually create garbage.

Reply:
I've mostly optimized away all the garbage here. I do an initial IV clone when calling the public `init` method, but that's the only time. I don't see the IV being cloned further by the `Cipher` classes that use `GCMParameterSpec`.

Reply:
Wow, I optimized it so much, both AES/GCM engines can't tell that the IV changes between invocations and throw an exception. Nice.

Reply:
I can't figure out how to get the engine to consider the IV as different, so I still have some clone calls happening, but much less waste than before.

Reply:
I looked into the internal implementation of the cipher engine, and it makes its own clone regardless of what we do, so this seems to be about as optimal as I can get it without digging into Bouncycastle internals or something similar.